src/null-safety/migration-guide.md

```diff
@@ -1,23 +1,453 @@
-[Dart blog]: https://medium.com/dartlang
+[null safety]: /null-safety
 
+![Screenshot of migration tool](/null-safety/migration-tool.png)
+_The interactive migration tool can simplify migration to null safety._
+
+
+## 1. Wait to migrate {#step1-wait}
+
+We strongly recommend migrating code in order, 
+with the leaves of the dependency graph being migrated first.
+For example, if package C depends on package B, which depends on package A,
+then A should be migrated to null safety first, then B, then C.
+
+![Illustration of C/B/A sentence](/null-safety/null-safety-migration-order.png){:width="454px"}<br>
+
+Although you _can_ migrate before your dependencies support null safety,
+you might have to change your code when your dependencies migrate.
+For example, if you predict that a function will take a nullable parameter but
+the package migrates it to be non-nullable,
+then passing a nullable argument becomes a compile error.
+
+This section tells you how to
+check your package's dependencies
+using the `dart pub outdated` command in null-safety mode.
+
+{{ site.alert.warn }}
+  These instructions assume your code is under **source control**,
+  so that you can easily undo any changes.
+{{ site.alert.end }}
+
+### Switch to a 2.12 beta release
+
+**Switch to the latest beta release**
+of either the Dart SDK or the Flutter SDK.
+How you get the latest beta release depends on whether
+you use the Flutter SDK:
+
+* If you use the Flutter SDK,
+  switch to the beta channel:
+
+  ```terminal
+$ flutter channel beta
+$ flutter upgrade
+```
+* Otherwise, download a beta release
+  from the [Dart SDK archive][].
+
+[Dart SDK archive]: /tools/sdk/archive#beta-channel
+
+{{ site.alert.warn }}
+  If you return to working on production code, remember to
+  **switch back to a stable release**
+  (for example, by running `flutter channel stable`).
+{{ site.alert.end }}
+
+
+### Check dependency status
+
+Get the migration state of your package's dependencies,
+using the following command:
+
+```terminal
+$ dart pub outdated --mode=null-safety
+```
+
+If the output says **all the packages fully support null safety**,
+then you can start migrating.
+Otherwise, use the **Resolvable** column to find
+null-safe releases, if they exist.
+
+{{ site.alert.info }}
+  **Why do all dependencies need to support null safety?**
+  When all of an app's direct dependencies support null safety,
+  you can _run the app_ with sound null safety.
+  When all the dev dependencies support null safety,
+  you can _run tests_ with sound null safety.
+  You might also need null-safe dev dependencies for other reasons,
+  such as code generation.
+{{ site.alert.end }}
+
+Here's an example of the output for a simple package.
+The green checkmarked version for each package supports null safety:
+
+![Output of dart pub outdated, showing that examples_util has no releases with null safety support, but pedantic and test do](/null-safety/pub-outdated-output.png)
```

Comment:
The image seems outdated? It doesn't show the null safe version of `examples_util` that have in step 2 below.

Also, I think it would be even better if we had both a direct and dev dependencies.

The test case I've been using is:

`dart create sampleapp`

Edit `pubspec.yaml` to:
```
environment:
  sdk: '>=2.8.1 <3.0.0'

dependencies:
  path: ^1.7.0
  process: 3.0.13

dev_dependencies:
  pedantic: ^1.9.0
```

Run `pub get`

Then `dart pub outdated --mode=null-safety` gives me:

```
direct dependencies:
path          ✗1.7.0   ✗1.7.0      ✓1.8.0-nullsafety.3   ✓1.8.0-nullsafety.3
process       ✗3.0.13  ✗3.0.13     ✓4.0.0-nullsafety.4   ✓4.0.0-nullsafety.4

dev_dependencies:
pedantic      ✗1.9.2   ✗1.9.2      ✓1.10.0-nullsafety.3  ✓1.10.0-nullsafety.3
```

Reply:
I was trying to illustrate the point the you can't migrate yet (step 1: wait for examples_util to be migrated). But apparently that isn't clear. Switching a different example sounds good to me; I'll work on that.